src/halmos/cheatcodes.py

```diff
@@ -216,6 +219,108 @@ def symbolic_storage(ex, arg, sevm, stack, step_id):
+        ex, arg, sevm, stack, step_id, contract_name, filename
+    )
+
+
+def create_calldata_contract(ex, arg, sevm, stack, step_id):
+    contract_name = name_of(extract_string_argument(arg, 0))
+    return create_calldata_generic(ex, arg, sevm, stack, step_id, contract_name)
+
+
+def create_calldata_generic(
+    ex, arg, sevm, stack, step_id, contract_name, filename=None
+):
+    """
+    Generate arbitrary symbolic calldata for the given contract.
+
+    Dynamic-array arguments are sized in the same way of regular test functions.
+
+    The contract is identified by its contract name and optional filename.
+    TODO: provide variants that require only the contract address.
+    """
+    contract_json = BuildOut().get_by_name(contract_name, filename)[0]
+
+    abi = contract_json["abi"]
+    methodIdentifiers = contract_json["methodIdentifiers"]
+
+    results = []
+
+    for funsig in methodIdentifiers:
+        funname = funsig.split("(")[0]
+        funselector = methodIdentifiers[funsig]
+        funinfo = FunctionInfo(funname, funsig, funselector)
+
+        dyn_param_size = DynamicParams()
+
+        calldata = ByteVec()
+        calldata.append(int(funselector, 16).to_bytes(4, "big"))
+        mk_calldata(
+            abi,
+            funinfo,
+            calldata,
+            dyn_param_size,
+            sevm.options,
+            new_symbol_id=ex.new_symbol_id,
+        )
+
+        calldata_lst = (
+            permutate_dyn_size(dyn_param_size, funselector, abi, funinfo, sevm, ex)
+            if dyn_param_size
+            else [calldata]
+        )
+
+        for calldata in calldata_lst:
+            data = calldata.unwrap()
+            if isinstance(data, bytes):
+                data = bytes_to_bv_value(data)
+
+            results.append(ByteVec(Concat(con(32), con(len(calldata)), data)))
+
+    return results
+
+
+def permutate_dyn_size(dyn_param_size, funselector, abi, funinfo, sevm, ex):
+    arrlen_lst = [{}]
+    for p_name, p_size, p_typ in dyn_param_size:
+        # TODO: provide cli flags to specify these values
+        new_size_options = (
+            [0, 1, 2]  # array lengths
+            if isinstance(p_typ, DynamicArrayType)
+            else [0, 32, 65, 1024]  # bytes/string lengths
+        )
+        if p_size not in new_size_options:
+            new_size_options.append(p_size)
+
+        # consider all size combinations
+        arrlen_lst = [
+            {**arrlen, p_name: new_size}
+            for arrlen in arrlen_lst
+            for new_size in new_size_options
+        ]
+
+    result = []
+    for arrlen in arrlen_lst:
+        calldata = ByteVec()
+        calldata.append(int(funselector, 16).to_bytes(4, "big"))
+        mk_calldata(
+            abi,
+            funinfo,
+            calldata,
+            DynamicParams(),
+            sevm.options,
+            arrlen=arrlen,
+            new_symbol_id=ex.new_symbol_id,
+        )
+        result.append(calldata)
+
+    return result
+
+
```

Comment:
note: the combination generation logic has an issue with duplication, as shown in HalmosCheatCode test. this will be addressed in a separate pr.

Reply:
fixed in https://github.com/a16z/halmos/pull/360/commits/36ecb35bdd86e86093ceb08912c9087c8e3dd365